scripts/split_repo_migration/git-migrate-history.sh

```diff
@@ -0,0 +1,173 @@
+then
+  echo "Pruning commits only including path: ${SOURCE_PATH}"
+  git filter-branch \  # rewrites history... unsafe
+    --prune-empty \
+    --subdirectory-filter "${SOURCE_PATH}"
+fi
+
+if [[ ! -z "${IGNORE_FOLDERS}" ]]
+then
+  echo "Ignoring folder: ${IGNORE_FOLDERS}"
+  mkdir -p "${WORKDIR}/filtered-source"
+  FOLDERS=$(echo ${IGNORE_FOLDERS} | tr "," " ")
+  # remove files/folders we don't want
+  FILTER="(rm -rf ${FOLDERS} || true)"
+  if [[ ! -z "${KEEP_FILES}" ]]
+  then
+    KEEP_FILES_SPACES=($(echo ${KEEP_FILES} | tr "," " "))
+    LAST_ELEMENT=$(( ${#KEEP_FILES_SPACES[@]} - 1 ))
+    KEEP_FILE_COMMANDS=""
+    for file in "${KEEP_FILES_SPACES[@]}"
+    do
+      if [[ $file == "${KEEP_FILES_SPACES[$LAST_ELEMENT]}" ]]
+      then
+        KEEP_FILE_COMMANDS+="git checkout -- ${file} 2>/dev/null || true"
+      else 
+        KEEP_FILE_COMMANDS+="git checkout -- ${file} 2>/dev/null || true; "
+      fi   
+    done
+    # restore files to keep, silence errors if the file doesn't exist
+    FILTER="${FILTER}; ${KEEP_FILE_COMMANDS}"
+  fi
+  git filter-branch \
+    --force \
+    --prune-empty \
+    --tree-filter "${FILTER}"
+fi
+
+DISTRIBUTION_NAME=$(jq -r '.distribution_name' .repo-metadata.json) # -r removes quotes around the name.
+TARGET_PATH="packages/${DISTRIBUTION_NAME}"
+
+# reorganize the filtered code into the desired target locations
+echo "Moving files to destination path: ${TARGET_PATH}"
+git filter-branch \
+  --force \
+  --prune-empty \
+  --tree-filter \
+    "shopt -s dotglob; mkdir -p ${WORKDIR}/migrated-source; mv * ${WORKDIR}/migrated-source; mkdir -p ${TARGET_PATH}; mv ${WORKDIR}/migrated-source/* ${TARGET_PATH}"
+
+# back to workdir
+popd
+
+# merge histories
+pushd $TARGET_REPO
+
+REMOTE="remote.${SOURCE_REPO}"
+git remote add --fetch ${REMOTE} ${WORKDIR}/source-repo
+git checkout -B "${BRANCH}"
+git merge --allow-unrelated-histories ${REMOTE}/main --no-edit
+
+echo "Success"
+
+popd # back to workdir
+
+# Do a diff between source code split repo and migrated code.
+git clone "git@github.com:${SOURCE_REPO}.git" source-repo-validation  # Not ideal to clone again.
+rm -rf source-repo-validation/.git  # That folder is not needed for validation.
+
+DIFF_FILE="${WORKDIR}/diff.txt"
+if diff -r "${TARGET_REPO}/${TARGET_PATH}" source-repo-validation > "${DIFF_FILE}" ; then
+  echo "No diff"
+else
+  echo "Diff non-empty. See ${DIFF_FILE}"
+  $EXIT 1
+fi
+
+pushd "${TARGET_REPO}"  # To target repo
+
+# For postprocessing of the batch migration script.
+mkdir -p owl-bot-staging/${DISTRIBUTION_NAME}/${DISTRIBUTION_NAME}
+touch owl-bot-staging/${DISTRIBUTION_NAME}/${DISTRIBUTION_NAME}/${DISTRIBUTION_NAME}.txt
+git add owl-bot-staging
+git commit -m "Trigger owlbot post-processor"
+
+git push -u origin "${BRANCH}" --force
+
+# create pull request
+if gh --help > /dev/null
+then
+  gh pr create --title "migrate code from ${SOURCE_REPO}"  --body "feat: Migrate ${SOURCE_REPO} into ${TARGET_PATH} (#${ISSUE_NUMBER})."
```

Comment:
Doesn't the conventional commit need to appear in the title? For my sample PR this morning, I wound up manually changing the title to `chore(migration): Import googleapis/python-ids into packages/google-cloud-ids` and that passed the presubmit checks. Should we try something like that?

Reply:
Yeah, I updated title. It will be something like:

```chore: Import googleapis/python-ids into packages/google-cloud-ids (#11104)

See #11104.
```